pyfr/solvers/base/elements.py

```diff
@@ -211,47 +214,88 @@ def _gen_pnorm_fpts(self):
-        jacop = np.rollaxis(self.basis.sbasis.jac_nodal_basis_at(pts), 2)
+        # Interpolation matrix to pts
+        M0 = self.basis.mbasis.nodal_basis_at(pts)
+
+        # Interpolate smats
+        smats = np.array([np.dot(M0, smat) for smat in self._smats])
+        smats = smats.reshape(self.ndims, npts, self.ndims, -1)
+
+        # Interpolate djacs
+        if retdets:
+            djacs = np.dot(M0, self._djacs)
+
+        return (smats, djacs) if retdets else smats
+
+    def _ploc_at_pts_np(self, pts):
+        op = self.basis.sbasis.nodal_basis_at(pts)
+
+        ploc = np.dot(op, self.eles.reshape(self.nspts, -1))
+        ploc = ploc.reshape(-1, self.neles, self.ndims).swapaxes(1, 2)
+
+        return ploc
+
+    def _smat_base(self):
+        # Metric basis with grid point (q<=p) or pseudo grid points (q>p)
+        mpts = self.basis.mpts
+        mbasis = self.basis.mbasis
+        self.nmpts = nmpts = len(mpts)
+
+        # Dimensions and number of elements
+        ndims = self.ndims
+        neles = self.neles
+
+        # Coordinate at pts
+        x = self._ploc_at_pts_np(mpts)
+
+        # Jacobian at pts
+        jacop = np.rollaxis(mbasis.jac_nodal_basis_at(mpts), 2)
+        jacop = jacop.reshape(-1, nmpts)
 
         # Cast as a matrix multiply and apply to eles
-        jac = np.dot(jacop.reshape(-1, nspts), self.eles.reshape(nspts, -1))
+        jac = np.dot(jacop, x.reshape(nmpts, -1))
 
         # Reshape (npts*ndims, neles*ndims) => (npts, ndims, neles, ndims)
-        jac = jac.reshape(npts, ndims, neles, ndims)
+        jac = jac.reshape(nmpts, ndims, ndims, neles)
 
         # Transpose to get (ndims, npts, ndims, neles)
-        return jac.transpose(3, 0, 1, 2)
+        jac = jac.transpose(2, 0, 1, 3)
 
-    def _get_smats(self, pts, retdets=False):
-        jac = self._get_jac_eles_at(pts)
         smats = np.empty_like(jac)
 
-        if self.ndims == 2:
+        if ndims == 2:
+            # Just compute smats whose order is q-1
             a, b, c, d = jac[0,:,0], jac[0,:,1], jac[1,:,0], jac[1,:,1]
 
             smats[0,:,0], smats[0,:,1] = d, -b
             smats[1,:,0], smats[1,:,1] = -c, a
 
-            if retdets:
-                djacs = a*d - b*c
+            self._djacs = a*d - b*c
         else:
-            # We note that J = [x0, x1, x2]
-            x0, x1, x2 = jac[:,:,0], jac[:,:,1], jac[:,:,2]
-
-            smats[0] = np.cross(x1, x2, axisa=0, axisb=0, axisc=1)
-            smats[1] = np.cross(x2, x0, axisa=0, axisb=0, axisc=1)
-            smats[2] = np.cross(x0, x1, axisa=0, axisb=0, axisc=1)
-
-            if retdets:
-                # Exploit the fact that det(J) = x0 . (x1 ^ x2)
-                djacs = np.einsum('ij...,ji...->j...', x0, smats[0])
-
-        return (smats, djacs) if retdets else smats
+            # Cpmpute x cross x_(chi)
+            jac = np.rollaxis(jac, 2)
+            tt = np.array([np.cross(x, dx, axisa=1, axisb=0, axisc=1)
```

Comment:
Make this a list then call `np.array(tt).reshape(...)` on the next line.  Should save us a line